docs/components/authentication.md

```diff
@@ -0,0 +1,185 @@
+
+``` php
+use Http\Message\Authentication\Bearer;
+
+$authentication = new Bearer('token');
+
+// Token can be set later as well
+$authentication->setToken('token');
+```
+
+
+### Chain
+
+This authentication method accepts one parameter: an array of other authentication methods.
+
+The idea behind this authentication method is that in some cases you might need to
+authenticate the request with multiple methods.
+
+For example it's a common practice to protect development APIs with Basic Auth and the regular token auth as well
+to protect the API from unnecessary processing.
+
+
+``` php
+use Http\Message\Authentication\Chain;
+
+$authenticationChain = [
+    new AuthenticationMethod1(),
+    new AuthenticationMethod2(),
+];
+
+$authentication = new Chain($authenticationChain);
+
+// Authentication chain can be modified later
+$authenticationChain = $authentication->getAuthenticationChain();
+
+array_pop($authenticationChain);
+
+$authentication->setAuthenticationChain($authenticationChain);
+
+$authentication->clearAuthenticationChain();
+
+$authentication->addAuthentication(new AuthenticationMethod3());
+```
+
+
+### Matching
+
+With this authentication method you can conditionally add authentication details to your request by passing a callable
+to it. When a request is passed, the callable is called and used as a boolean value in order to decide whether
+the request should be authenticated or not.
+It also accepts an authentication method instance which does the actual authentication when the condition is
+fulfilled.
+
+For example a common use case is to authenticate requests sent to certain paths.
+
+
+``` php
+use Http\Message\Authentication\Mathing;
+
+$authentication = new Mathing(new AuthenticationMethod1(), function () { return true; });
+
+// There are also getters with the appropriate method names
+$authentication->setAuthentication(new AuthenticationMethod2());
+$authentication->setMatcher(function () { return false; });
+```
+
+
+In order to ease creating matchers for URLs/paths, there is a static factory method for this purpose.
+The first argument is an authentication method, the second is a regexp to match against the URL.
+
+
+``` php
+use Http\Message\Authentication\Mathing;
+
+$authentication = Matching::createUrlMatcher(new AuthenticationMethod(), '\/api');
+```
+
+
+### Query Params
+
+Add authentication details as URL Query params:
+
+`http://api.example.com/endpoint?access_token=9zh987g86fg87gh978hg9g79`
+
+
+``` php
+use Http\Authentication\QueryParams;
+
+$authentication = new QueryParams([
+    'access_token' => '9zh987g86fg87gh978hg9g79',
+]);
+```
+
+!!! warning "Warning:"
+    Using query parameters for authentication is not safe.
+    Only use it when absolutely necessary.
+
+
+### WSSE
+
+This method implements [WSSE Authentication](http://www.xml.com/pub/a/2003/12/17/dive.html).
+Just like Basic Auth, it also accepts a username-password pair and exposes the same methods as well.
```

Comment:
i like the link to the standard. we should do that for all methods. not sure if the spec is the right place, or wikipedia might be easier to read. but we should link them to something.

also, we might order these methods semantically rather than alphabetically, and put this next to basic auth.